server/src/main/java/org/elasticsearch/bootstrap/JNAFalloc.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.bootstrap;
+
+import com.sun.jna.Native;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.Constants;
+import org.elasticsearch.common.Nullable;
+
+/**
+ * System specific wrappers of the fallocate system call via JNA for Linux and OSX.
+ */
+abstract class JNAFalloc {
+
+    private static final Logger logger = LogManager.getLogger(JNAFalloc.class);
+
+    public abstract int fallocate(int fd, long offset, long length);
+
+    @Nullable
+    public static JNAFalloc falloc() {
+        try {
+            if (Constants.MAC_OS_X) {
+                return OSX.INSTANCE;
+            } else if (Constants.LINUX) {
+                return Linux.INSTANCE;
+            }
+        } catch (Throwable t) {
+            // just debug, already logged in the specific classes
+            logger.debug("failed to load JNA native falloc implementation", t);
+        }
+        return null;
+    }
+
+    private static class Linux extends JNAFalloc {
+
+        static final Linux INSTANCE = new Linux();
+
+        static {
+            try {
+                Native.register("c");
+            } catch (UnsatisfiedLinkError e) {
+                logger.warn("unable to link C library. native (falloc) will be disabled.", e);
+            }
+        }
+
+        @Override
+        public int fallocate(int fd, long offset, long length) {
+            final int res = fallocate(fd, 0, offset, length);
+            return res == 0 ? 0 : Native.getLastError();
+        }
+
+        // TODO: Bind POSIX fallocate as well to support non-Linux? (this would only apply to OSX in practice?)
```

Comment:
Is this comment up to date given the OSX definition below?

Reply:
Ah thanks for spotting, it's not :) though ... this needed another change anyway, turns out this doesn't work on every Mac still. So I had to adjust some more.